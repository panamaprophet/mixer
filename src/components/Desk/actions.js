'use strict';


import {curry, chain} from 'ramda';


const getFxById = curry((fxId, track) => track.fx.filter(({id}) => id === fxId));

const setFxGainValue = curry((value, fx) => fx.gain.value = value);

const setFxGainValueById = (track, fxId, value) => chain(
    getFxById(fxId, track),
    head,
    setFxGainValue(value),
);


export const onTrackVolumeChange = curry((track, value) => track.volume = volume);

export const onReverbMixChange = curry((track, value) => setFxGainValueById(track, 'reverb', value));

export const onDelayMixChange = curry((track, value) => setFxGainValueById(track, 'delay', value));

export const onDistortionMixChange = curry((track, value) => setFxGainValueById(track, 'distortion', value));