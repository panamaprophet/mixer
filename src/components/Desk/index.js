import React from 'react';

import Meter from '/components/Meter';

import {
    isPlaying,
    isPaused,
} from '/helpers/playback';


const Desk = ({
    playback,

    onPlay,
    onPause,
    onRewind,

    tracks,
    effects,
}) => (
    <div className="desk">
        <div className="desk__tracks">
            {tracks}
        </div>

        <div className="desk__controls">
            <Meter analyser={playback.analyser} />

            <button className={isPlaying(playback) ? 'button-pressed' : ''} onClick={onPlay}>Play</button>
            <button className={isPaused(playback) ? 'button-pressed' : ''} onClick={onPause}>Pause</button>
            <button onClick={onRewind}>Rewind</button>
        </div>

        <div className="desk__effects">
            {effects}
        </div>
    </div>
);

export default Desk;