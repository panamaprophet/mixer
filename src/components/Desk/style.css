.desk {
    font-size: .8em;

    position: relative;

    display: flex;
    flex-flow: row wrap;
    flex-grow: 1;
    justify-content: space-between;

    width: 100%;
    height: 620px;
    padding: 0;

    user-select: none;
    -webkit-user-select: none;

    border-radius: 4px;
    background: none;
}

.controlsContainer {
    flex-grow: 1;
    flex-shrink: 1;

    max-width: 35%;
    padding: 0;

    border-radius: 4px;
    background: none;
}

.tracks {
    display: flex;
    overflow: hidden;
    flex-flow: row nowrap;
    flex-grow: 1;
    flex-shrink: 1;

    max-width: 63%;
    height: 100%;

    border-radius: 4px;
    box-shadow: 0 0 3px #000;
}

.controls {
    font-size: 0;

    display: flex;
    flex-flow: row wrap;

    margin-bottom: 2px;
    padding: 16px;

    border-radius: 4px 4px 0 0;
    background: #c4cace; /* url("i/metal-texture-small.jpg"); */
    background-size: contain;
    box-shadow: 0 0 3px #000;
}

.button {
    font-size: .8rem;
    font-weight: bold;

    flex-grow: 1;

    margin-right: 1px;
    padding: 8px 16px;

    cursor: pointer;
    transition: all 80ms ease-in-out;
    text-transform: capitalize;

    border: none;
    outline: none;
    background: #eee;
    box-shadow: inset 0 0 0 1px rgba(0,0,0,.2), inset 0 0 1px 2px rgba(255,255,255,.9), inset 0 -6px 5px rgba(0,0,0,.1), 0 6px 7px rgba(0,0,0,.3), 0 4px 1px rgba(0,0,0,.5);
    text-shadow: 1px 0 0 #efefef;
}

.button:hover {
    box-shadow: inset 0 0 0 1px rgba(0,0,0,.2), inset 0 0 1px 2px rgba(255,255,255,.9), inset 0 -10px 15px rgba(0,0,0,.1), 0 6px 7px rgba(0,0,0,.3), 0 4px 1px rgba(0,0,0,.5);
}

.button:active,
.isButtonPressed {
    transform: translateY(2px);

    box-shadow: inset 0 0 0 1px rgba(0,0,0,.18), inset 0 0 1px 2px rgba(255,255,255,.5), inset 0 -6px 5px rgba(0,0,0,.1), 0 6px 7px rgba(0,0,0,.3), 0 2px 1px rgba(0,0,0,.5);
}

.button[disabled] {
    color: grey;
}

.effects {
    font-size: .8rem;

    display: flex;
    flex-flow: column wrap;
    justify-content: space-between;
}


@media only screen and (max-width: 375px) {
    .tracks,
    .controlsContainer
    {
        max-width: 100%;
        margin-bottom: 8px;
    }
}
