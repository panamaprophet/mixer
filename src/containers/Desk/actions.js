export const play = dispatch => dispatch({type: 'CONTROLS_PLAY'});

export const pause = dispatch => dispatch({type: 'CONTROLS_PAUSE'});

export const rewind = dispatch => dispatch({type: 'CONTROLS_REWIND'});