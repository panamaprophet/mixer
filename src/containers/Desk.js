import React, {useContext} from 'react';

import Desk from '/components/Desk';
import Context from '/containers/Context';
import Track from '/containers/Track';

const DeskContainer = ({
    tracks,
    effects,
    controls,
}) => {
    const dispatch = useContext(Context);
    const {analyser} = controls;

    const play = () => dispatch({type: 'CONTROLS_PLAY'});
    const pause = () => dispatch({type: 'CONTROLS_PAUSE'});
    const rewind = () => dispatch({type: 'CONTROLS_REWIND'});

    const Tracks = tracks.map(track => (<Track {...track} key={track.id} />));
    const Effects = null;

    return (
        <Desk 
            onPlay={play}
            onPause={pause}
            onRewind={rewind}

            tracks={Tracks}
            effects={Effects}
            analyser={analyser}
        >
        </Desk>
    );
}

export default DeskContainer;