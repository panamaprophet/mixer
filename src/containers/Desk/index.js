import React, {useContext} from 'react';

import Desk from '/components/Desk';
import Effect from '/containers/Effect';
import Context from '/containers/Context';
import Track from '/containers/Track';

import {
    play,
    pause,
    rewind,
} from './actions';

const DeskContainer = ({
    tracks,
    effects,
    playback,
}) => {
    const dispatch = useContext(Context);

    const Tracks = tracks.map(track => (<Track {...track} key={track.id} />));
    const Effects = effects.map(effect => (<Effect {...effect} key={effect.id} />));

    return (
        <Desk 
            onPlay={() => play(dispatch)}
            onPause={() => pause(dispatch)}
            onRewind={() => rewind(dispatch)}

            tracks={Tracks}
            effects={Effects}
            playback={playback}
        >
        </Desk>
    );
}

export default DeskContainer;