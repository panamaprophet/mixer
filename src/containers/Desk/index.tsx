import React, {useContext} from 'react';

import Desk from '/components/Desk';
import Effect from '/containers/Effect';
import Context from '/containers/Context';
import Track from '/containers/Track';

import {
    play,
    pause,
    rewind,
} from '/store/actions';

type Props = {
    tracks: any,
    effects: any,
    playback: any,
}


const DeskContainer = ({
    tracks,
    effects,
    playback,
}: Props) => {
    const dispatch = useContext(Context);

    const Tracks = tracks.map(track => (<Track {...track} key={track.id} />));
    const Effects = effects.map(effect => (<Effect {...effect} key={effect.id} />));

    return (
        <Desk
            onPlay={() => play(dispatch)}
            onPause={() => pause(dispatch)}
            onRewind={() => rewind(dispatch)}

            tracks={Tracks}
            effects={Effects}
            playback={playback}
        >
        </Desk>
    );
}

export default DeskContainer;