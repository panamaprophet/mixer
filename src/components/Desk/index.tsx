import React from 'react';
import classnames from 'classnames';
import {Meter} from '/components/Meter/index';
import {Icon} from '/components/Icon/index';
import {isPlaying, isPaused, Playback, isActive} from '/helpers/playback';
import style from './style.css';


type Props = {
    playback: Playback,
    onPlay: () => void,
    onPause: () => void,
    onRewind: () => void,
    tracks: React.ReactNode[],
    sends: React.ReactNode[],
}


export const Desk: React.FC<Props> = ({
    playback,
    tracks = [],
    sends = [],
    onPlay,
    onPause,
    onRewind,
}: Props) => {
    const btnClassNames = (isButtonPressed: boolean) => classnames(
        style.control,
        style.button,
        isButtonPressed && style.isButtonPressed,
    );

    const isDisabled = isActive(playback) === false;

    return (
        <div className={style.desk}>
            <div className={style.tracks}>
                {tracks}
            </div>

            <div className={style.controlsContainer}>
                <div className={style.controls}>
                    {playback.analyser && <Meter analyser={playback.analyser} />}

                    <button className={btnClassNames(isPlaying(playback))} onClick={onPlay} disabled={isDisabled}>
                        <Icon type="play" />
                        Play
                    </button>

                    <button className={btnClassNames(isPaused(playback))} onClick={onPause} disabled={isDisabled}>
                        <Icon type="pause" />
                        Pause
                    </button>

                    <button className={btnClassNames(false)} onClick={onRewind} disabled={isDisabled}>
                        <Icon type="rewind" />
                        Rewind
                    </button>
                </div>

                <div className={style.sends}>
                    {sends}
                </div>
            </div>
        </div>
    );
};