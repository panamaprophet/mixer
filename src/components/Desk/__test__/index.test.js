import React from 'react';
import {shallow} from 'enzyme';

import Desk from '../';

import {
    PLAYBACK_STATUS,
} from '/constants';

const playbackMock = {
    analyser: {},
    status: PLAYBACK_STATUS.PAUSED,
    currentPosition: 0,
};

const PLAY_BUTTON = 0;
const PAUSE_BUTTON = 1;
const REWIND_BUTTON = 2;


describe('<Desk />', () => {
    it('should renders without any errors', () => {
        const wrapper = shallow(
            <Desk playback={{
                ...playbackMock,
            }} />
        );

        expect(wrapper).toMatchSnapshot();
    });

    describe('playback buttons', () => {
        it('should be disabled if mixdesk is not active', () => {
            const wrapper = shallow(
                <Desk playback={{
                    ...playbackMock,
                    status: PLAYBACK_STATUS.NOT_SET,
                }} />
            );

            const result = wrapper.find('button[disabled]').length;

            expect(result).toBe(3);
        });
    });

    describe('play button', () => {
        it('should call onPlay on press', () => {
            const onPlay = jest.fn();
            const wrapper = shallow(<Desk playback={{
                ...playbackMock,
            }} onPlay={onPlay} />);

            wrapper.find('button').at(PLAY_BUTTON).simulate('click');

            expect(onPlay).toBeCalled();
        });

        it('should have active class on playing', () => {
            const wrapper = shallow(<Desk playback={{
                ...playbackMock,
                status: PLAYBACK_STATUS.PLAYING,
            }} />);

            const result = wrapper.find('button').at(PLAY_BUTTON).hasClass('isButtonPressed');

            expect(result).toBe(true);
        });
    });

    describe('pause button', () => {
        it('should call onPause on press', () => {
            const onPause = jest.fn();
            const wrapper = shallow(<Desk playback={{
                ...playbackMock,
            }} onPause={onPause} />);

            wrapper.find('button').at(PAUSE_BUTTON).simulate('click');

            expect(onPause).toBeCalled();
        });

        it('should have active class on pause', () => {
            const wrapper = shallow(<Desk playback={{
                ...playbackMock,
                status: PLAYBACK_STATUS.PAUSED,
                currentPosition: 420,
            }} />);

            const result = wrapper.find('button').at(PAUSE_BUTTON).hasClass('isButtonPressed');

            expect(result).toBe(true);
        });
    });

    describe('rewind button', () => {
        it('should call onRewind on press', () => {
            const onRewind = jest.fn();
            const wrapper = shallow(<Desk playback={{
                ...playbackMock,
            }} onRewind={onRewind} />);

            wrapper.find('button').at(REWIND_BUTTON).simulate('click');

            expect(onRewind).toBeCalled();
        });
    });
});