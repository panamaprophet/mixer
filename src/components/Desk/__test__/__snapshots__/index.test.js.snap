// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Desk /> should renders without any errors 1`] = `
<div
  className="desk"
>
  <div
    className="tracks"
  />
  <div
    className="controlsContainer"
  >
    <div
      className="controls"
    >
      <Meter
        analyser={Object {}}
      />
      <button
        className="control button"
        disabled={false}
        onClick={[Function]}
      >
        <Icon
          type="play"
        />
        Play
      </button>
      <button
        className="control button isButtonPressed"
        disabled={false}
        onClick={[Function]}
      >
        <Icon
          type="pause"
        />
        Pause
      </button>
      <button
        className="control button"
        disabled={false}
        onClick={[Function]}
      >
        <Icon
          type="rewind"
        />
        Rewind
      </button>
    </div>
    <div
      className="effects"
    />
  </div>
</div>
`;
